test/run_tests.py

```diff
@@ -0,0 +1,662 @@
+                                                                ref_content, ref_file, overwrite, epsilon, fuzzy_compare=fuzzy_compare)
+
+            if are_different:
+                message = "Diff not OK, {}".format((reason))
+            else:
+                message = "Diff OK, {}".format((reason))
+
+            checks[output_file] = {
+                "success": are_different == False,
+                "message": message,
+                "diff": diff
+            }
+    except:
+        completed_tests.report_completion(id, False)
+        raise
+
+    success = all(check["success"] == True for name, check in checks.items())
+    completed_tests.report_completion(id, success)
+
+    return (id, {
+        "result": Result.SUCCESS if success else Result.FAIL,
+        "checks": checks
+    })
+
+
+class Completion():
+    def __init__(self):
+        self.lock = threading.Lock()
+        self.condition = threading.Condition(self.lock)
+        self.completed = dict()
+
+    def report_completion(self, id, success):
+        self.lock.acquire()
+        self.completed[id] = success
+        self.condition.notify_all()
+        self.lock.release()
+
+    def wait_for_completion_get_success(self, id):
+        def is_complete():
+            return id in self.completed
+        self.lock.acquire()
+        if not is_complete():
+            self.condition.wait_for(is_complete)
+        success = self.completed[id]
+        self.lock.release()
+        return success
+
+
+def create_test_dir(test_id, input_files, test_base_dir, test_ref_dir, dependencies=None):
+    test_working_dir = Path(test_base_dir).joinpath(
+        "test_{}".format((test_id)))
+    Path(test_working_dir).mkdir(parents=True, exist_ok=True)
+
+    # Required as workaround until #2686 is fixed.
+    Path(test_working_dir.joinpath("models")).mkdir(
+        parents=True, exist_ok=True)
+
+    for file in input_files:
+        file_to_copy = None
+        search_paths = [Path(test_ref_dir).joinpath(file)]
+        if dependencies is not None:
+            search_paths.extend([Path(test_base_dir).joinpath(
+                "test_{}".format((x)), file) for x in dependencies])
+        for search_path in search_paths:
+            if search_path.exists() and not search_path.is_dir():
+                file_to_copy = search_path
+                break
+
+        if file_to_copy is None:
+            raise ValueError(
+                "{} couldn't be found for test {}".format((file), (test_id)))
+
+        test_dest_file = Path(test_working_dir).joinpath(file)
+        Path(test_dest_file.parent).mkdir(parents=True, exist_ok=True)
+        # We always want to replace this file in case it is the output of another test
+        if test_dest_file.exists():
+            test_dest_file.unlink()
+        shutil.copyfile(str(file_to_copy), str(test_dest_file))
+    return test_working_dir
+
+
+def main():
```

Comment:
would be nice if you could split main in to functions to have some more semantics i.e.

setup_arguments()
find_vw_binary()
...